internal/venice-common/src/main/java/com/linkedin/venice/writer/VeniceWriter.java

```diff
@@ -386,79 +400,56 @@ public void close(boolean gracefulClose) {
       long startTime = System.currentTimeMillis();
       logger.info("Closing VeniceWriter for topic: {}", topicName);
       try {
-        // If {@link #broadcastEndOfPush(Map)} was already called, the {@link #endAllSegments(boolean)}
-        // will not do anything (it's idempotent). Segments should not be ended if there are still data missing.
+        // try to end all segments before closing the producer
         if (gracefulClose) {
-          endAllSegments(true);
+          try {
+            endAllSegments().get(Math.max(100, closeTimeOutInMs / 2), TimeUnit.MILLISECONDS);
+          } catch (Exception e) {
+            logger.warn("Swallowed an exception while trying to end all segments for topic: {}", topicName, e);
+          }
         }
-        // DO NOT call the {@link #PubSubProducerAdapter.close(int) version from here.}
-        // For non-shared producer mode gracefulClose will flush the producer
-
-        producerAdapter.close(topicName, closeTimeOutInMs, gracefulClose);
+        getProducerAdapter().close(topicName, closeTimeOutInMs, gracefulClose);
         OPEN_VENICE_WRITER_COUNT.decrementAndGet();
       } catch (Exception e) {
         logger.warn("Swallowed an exception while trying to close the VeniceWriter for topic: {}", topicName, e);
         VENICE_WRITER_CLOSE_FAILED_COUNT.incrementAndGet();
       }
-      threadPoolExecutor.shutdown();
+      getThreadPoolExecutor().shutdown();
```

Comment:
Wrap inside `ifPresent` rather than doing a late initialization just for the sake of tearing it down right after...

Reply:
Good catch! Thanks